### PR TITLE
Don't erase event timezones

### LINKED ISSUE #650: Editing an event from Google Calendar changes the time by the UTC offset
## What Happened

I had a calendar event for 2 PM MST (UTC-7). I open it to edit, and the edit dialog still shows 2 PM. I changed the event from one day to another, leaving the time set to 2 PM. When I save the event, Calendar saves it in UTC (therefor changing it by the UTC offset, so 2 PM UTC = 7 AM MST) instead of the original/current time zone.

## Expected Behavior

Calendar would retain the timezone, and likely show the time zone any time a time is shown for clarity.

## Steps to Reproduce

1. Create a calendar event in Google Calendar with a timezone associated
2. Edit the event in elementary Calendar, e.g. changing the date but _not_ the time
3. See that the time is changed by the UTC offset

## Platform Information

elementary OS 6 Early Access with Calendar from master

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/96500668-editing-an-event-from-google-calendar-changes-the-time-by-the-utc-offset?utm_campaign=plugin&utm_content=tracker%2F60388692&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F60388692&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mcclurgm**
Fixes #650 
~~Or at least part of it. We no longer set the timezones to floating, but this doesn't (yet) preserve the original timezone either, instead changing everything to the system timezone. It also doesn't add any affordance in the UI for time zones.~~
This now preserves the original timezone and displays timezone information in the dialog UI.

The issue was that we tried to set the timezone before we told libical that the time isn't `DATE` type. Libical just doesn't change the timezone in that case. So this reorganizes that code to fix it, and also adds tests to make sure that doesn't happen again! While I'm here, I also added support for floating timezones (outside of the implicitly floating `DATE` types) by changing the timezone argument behavior.

Still to do:
- [x] Preserve the original timezone where possible
- [x] Figure out how to use tests with a TimeManager 😠️

### CODE REVIEW

**Comment by mcclurgm** [comment]
Alright, I think this is ready to go. I messed with the TimeManager a bit to separate it from DBus in a testing mode, which you may or may not want. I'd also appreciate some help testing that the new logic to preserve the original timezone works as intended, since there seem to be some pitfalls there. (You can check and modify timezones in Evolution.)

**Comment by mcclurgm** [comment]
Ok, there's one thing that I want to discuss here. I just went ahead and implemented converting the timezone back to the original, but thinking back on it I'm not convinced that was a great idea. My hope was to make our lack of timezone support invisible: it displays in the system timezone at all times in Calendar, but if you go to another app that does support timezones, the original timezone info would be intact. Ideally, this would all work together: there could be no issues with this conversion, and it would be truly transparent. However, it's possible (I don't understand time calculations well enough) that this could cause some discrepancies. There's also the question of whether we want to do these implicit conversions at all, or if we should save the time into the event in exactly the way it's presented to the user. Does anyone have thoughts on this? If we wanted to change the saved timezone to the system timezone, that would be a very quick change to make.

**Comment by mcclurgm** [comment]
Sorry to spam the comments in my own PR, but I also realized another potential issue with the current method: we'd also have to deal with the recurrence rules, which can have time components. We only display the date, but that's a potential issue I hadn't considered.

Overall, it will be more work (including design) to get everything to work in the original timezone instead of local, but could prevent some odd edge cases.

**Comment by mcclurgm** [comment]
Ok, I think this is ready to go. Sorry it's taken me so long to get it in this state.

I would ideally like to test the editor behavior to make sure things are working properly, but that's super complicated since it depends on multiple EDS objects that I don't know how I would mock.

I would appreciate input on two things in particular:

- As discussed on Slack, I convert events with multiple time zones (different for start and end) to the same time zone. This change is permanent (and destructive of the multiple-time zone data) when you save the event. Does this sound ok? Also, I made the arbitrary decision to convert the end time to the same time zone as the start time, instead of the other way around.
- The timezone display could use some design work. That's just what I threw together as a proof of concept, but I think it could be a lot nicer in the final version.

File: core/Tests/util-tests.vala [hunk 1]
```diff
@@ -314,20 +441,24 @@ void add_daterange_tests () {
...
     print ("\n");
     var original_tz = Environment.get_variable ("TZ");
     Environment.set_variable ("TZ", "America/Chicago", true);
     print ("Setting $TZ environment variable: " + Environment.get_variable ("TZ") + "\n");
+    ICal.Timezone tz = ICal.Timezone.get_builtin_timezone ("America/Chicago");
+    Calendar.TimeManager.setup_test (tz);
+    print ("Setting up TimeManager with system timezone America/Chicago");
     print ("Starting utils tests:\n");
 
     Test.init (ref args);
```
Comment (marbetschar) [inline core/Tests/util-tests.vala:457]: mhm, seems like this is similar to the `setUp` from JUnit, right? So there should probably be some way to access test-related variables/settings?

**Review by marbetschar — CHANGES_REQUESTED** [review]
This is pretty dope - thanks for digging into the Time Zone Limbo!! Also, it's really great to see you backed all of this with tests!

Unfortunately I'm not able to verify this from a functional perspective, since I don't have to deal with multiple timezones, so I just looked at the code and left some nitpicking comments :)

File: core/Services/Calendar/Util/DateTime.vala [hunk 2]
```diff
@@ -43,49 +43,67 @@ namespace Calendar.Util {
...
+#else
+            result.zone = timezone;
+#endif
+
+            // Set the time with the updated time zone
 #if E_CAL_2_0
-            result.set_is_date (false);
             result.set_time (time_local.get_hour (), time_local.get_minute (), time_local.get_second ());
 #else
             result._is_date = 0;
```
[the remark this answers is not in the document]
Comment (mcclurgm) [reply core/Services/Calendar/Util/DateTime.vala:103]: I expect that's an oversight. Good catch.

File: core/Services/Calendar/TimeManager.vala [hunk 3]
```diff
@@ -114,4 +114,20 @@ public class Calendar.TimeManager : Object {
...
+            warning ("Resetting default TimeManager to new testing instance");
+        }
+
+        instance = new TimeManager.for_testing (system_timezone);
+        return instance;
+    }
+
+    private TimeManager.for_testing (ICal.Timezone system_timezone) {
+        this.system_timezone = system_timezone;
+    }
```
[the remark this answers is not in the document]
Comment (mcclurgm) [reply core/Services/Calendar/TimeManager.vala:132]: The difficulty here is that `TimeManager` (as designed, I suppose that could change) is a singleton. It's used within other functions using the `get_default` function, so just generating a `TimeManager` instance in a test won't mean that the new instance is used by any other code. Does that make sense? 

If we want to do something like you suggest here and not overwrite the default instance, I'd have to refactor all the other code that uses `TimeManager`.

File: core/Tests/util-tests.vala [hunk 4]
```diff
@@ -314,20 +441,24 @@ void add_daterange_tests () {
...
+    Test.add_func ("/Utils/DateTime/datetimes_to_icaltime", test_datetimes_to_icaltime);
 }
 
 int main (string[] args) {
     print ("\n");
     var original_tz = Environment.get_variable ("TZ");
     Environment.set_variable ("TZ", "America/Chicago", true);
     print ("Setting $TZ environment variable: " + Environment.get_variable ("TZ") + "\n");
+    ICal.Timezone tz = ICal.Timezone.get_builtin_timezone ("America/Chicago");
+    Calendar.TimeManager.setup_test (tz);
```
[the remark this answers is not in the document]
Comment (mcclurgm) [reply core/Tests/util-tests.vala:453]: I'm not familiar with JUnit or GLib.Test, so I'll have to do some digging. If setup/teardown functions exist in GLib (which is _very_ sparse, so I don't trust that they do), then that's a good idea and I would like to use them.

File: core/Services/Calendar/TimeManager.vala (hunk 3 above)
[the remark this answers is not in the document]
Comment (marbetschar) [reply core/Services/Calendar/TimeManager.vala:132]: I see. Alright, I guess this can be done in a subsequent PR then.

File: core/Tests/util-tests.vala (hunk 4 above)
[the remark this answers is not in the document]
Comment (marbetschar) [reply core/Tests/util-tests.vala:453]: I'm not very familiar with GLib.Test either :) Also this does not seem to be a big improvement as long as the TimeManager is not refactored, I guess this is not a huge deal for now.

File: core/Tests/util-tests.vala (hunk 1 above)
Comment (mcclurgm) [reply core/Tests/util-tests.vala:457]: By my understanding, `Test.init` is used to initialize a test suite, not a single test. 

There does appear to be some setup/teardown functionality in the `add` function, but it's quite opaque and under-documented. I'll have to see if I can get it to work in Vala. It seems to depend on some weird parts of C.

File: core/Services/Calendar/TimeManager.vala (hunk 3 above)
[the remark this answers is not in the document]
Comment (mcclurgm) [reply core/Services/Calendar/TimeManager.vala:132]: Alternatively, if you can think of a way to make TimeManager not a singleton, that would be great. Most of it doesn't require being a singleton really. The only issue is that since #629 we store the system time zone in it for speed. We have to share that data somehow (or get it to the places where it's relevant) without re-calculating it every time, and I don't have any great ideas for how to do that without some universal state like a singleton.

File: core/Tests/util-tests.vala (hunk 4 above)
[the remark this answers is not in the document]
Comment (mcclurgm) [reply core/Tests/util-tests.vala:453]: Ok. If this PR gets merged in this state then I'll open an issue for it.

**Comment by mcclurgm** [comment]
Thanks for the suggestions, especially about the tests. It definitely feels a little strange, but it's the best I could come up with.

As for testing multiple timezones: if you want, you can create events in Evolution with any time zone you choose. They'll transfer over to elementary, so then you can make sure they work. It may not be the same as working with a full calendar, but it is something 🙂

**Review by tintou — APPROVED** [review]
Thanks for dealing with this, it looks good to me :rocket:

**Comment by marbetschar** [comment]
@mcclurgm did the functional testing and it works as expected. One small thing I came across is when adding new events: There the timezone information is not displayed at my end - it only appears after saving the event and editing it. Can we either display the system timezone there - or remove the whole timezone information completely?

The following screenshot shows how creating a new event looks like on my end:

Screenshot from 2021-09-16 08-19-55

**Comment by marbetschar** [comment]
@mcclurgm another thing which I came across is more from a design point of view: Since the timezone information is not editable anyways, any chance we can move it below the "All day" switch? Something like this:

Screenshot from 2021-09-16 08-25-33

**Comment by mcclurgm** [comment]
Thanks all. Marco, I'll have to dig in to creating new events. That's a big oversight on my part, I've only tested with pre-existing events. I'll get to work on that.

As for moving the time zone, that sounds like a nice idea, but I'm not sure how well it would work in practice. Not all time zones are that short, especially in Windows format. (We can't use those now, but I have a local branch that's working on it.) In the future I'd also like to make a selector, which I don't know it would look squeezed in over there. 
Perhaps since we now have 2 controls that affect both times, we could move both the all day switch and the time zone to a new row?
I'll keep experimenting.

**Comment by mcclurgm** [comment]
I just pushed a commit to fix the empty timezone issue. Turns out that to make that work nicely, I also made the code a little cleaner and set the stage for if we want to add a time zone selector, so that's nice.

**Review by marbetschar — APPROVED** [review]
Awesome, thank you! 🚀
